engine/engine.go

```diff
@@ -211,14 +209,43 @@ func (e Engine) launchResolution(publicID string, async bool, sm *semaphore.Weig
 
 	res.Values.SetConfig(e.config)
 
-	// all ready, run remaining steps
-
-	utask.AcquireExecutionSlot()
-
+	// check if all resources are available before starting the resolution
+	// first, check if we have a custom semaphore, for example, a semaphore that limits the concurrent execution of tasks recovery from a crashed instance.
+	// This semaphore needs to go first, because it will always be smaller than the global execution pool.
 	if sm != nil {
-		sm.Acquire(context.Background(), 1)
+		if err := sm.Acquire(shutdownCtx, 1); err != nil {
+			debugLogger.Debugf("Engine: launchResolution() %s acquire resource: instance is shutting down", res.PublicID)
+			return nil, errors.New("instance is shutting down")
+		}
+	}
+	// second, check if we have a resource limit on the current template
+	// template could be completely deactivated as a "dead resource". If it's the case, we need to exit because
+	// the limit won't change until next instance's reboot.
+	if acquiredErr := utask.AcquireResource(shutdownCtx, "template:"+t.TemplateName); acquiredErr != nil {
```

Comment:
You dont specify why you'd want to pass the shutdown context to the semaphore acquire calls?

From the doc: ```blocking until resources are available or ctx is done. ... If ctx is already done, Acquire may still succeed without blocking. ```

Wouldnt that possibly trigger weird storms in the seconds before shutdown, with all semaphores suddenly becoming non-blocking ?

Reply:
I'm giving the shutdownCtx to the semaphore acquire calls to release the "sleep" process, and then let the task be released.

> Wouldnt that possibly trigger weird storms in the seconds before shutdown, with all semaphores suddenly becoming non-blocking ?

It's possible that the semaphore succeed yes, but it won't go that far, as we check the shutdownCtx in the `runAvailableSteps` functions, meaning that no step will be evaluated. We will just give the task a good closing, otherwise, we would keep all those tasks in a `RUNNING` state, meanwhile it were only locked against the semaphore.